### PR TITLE
Add MetadataAnnotations class for user-facing handling of annotations

### PR DESCRIPTION
**Pull request by kbolashev**
Scope of this PR:
- Add a new dependency: https://github.com/DagsHub/dagshub-annotation-converter/
- Plus additionally add ultralytics into dev dependencies
- Add `MetadataAnnotations` class. Objects of this class will get created for any annotation field in the QueryResult, so you can use them like this:
```python
dp = ds.all()[0]
anns: MetadataAnnotations = dp["exported_annotations"]
anns.add_image_bbox("person", 0.1, 0.1, 0.1, 0.1)
anns.add_image_bbox("cat", 0.2, 0.2, 0.1, 0.1)
dp.save()
```
- Autoload annotations on `ds.all()` and `ds.head()`
- Add an ability to turn off this autoload functionality as `load_documents` and `load_annotations` kwargs in these functions

### CODE REVIEW

**Review by simonlsk — COMMENTED** [review]
All very cool features, added some comments and questions. 
Thanks!

File: dagshub/data_engine/model/query_result.py [hunk 1]
```diff
@@ -655,17 +683,16 @@ def document_fields(self) -> List[str]:
...
+            - All annotation fields
         """
-        if len(self.document_fields) > 0:
-            log_message(f"Downloading document fields {self.document_fields}...")
-            self.get_blob_fields(*self.document_fields, load_into_memory=True)
-            # Convert them to strings
-            for dp in self:
-                for f in self.document_fields:
-                    if f in dp.metadata:
-                        dp.metadata[f] = dp.metadata[f].decode("utf-8")
```
[the remark this answers is not in the document]
Comment (kbolashev) [reply dagshub/data_engine/model/query_result.py:671]: Moved to `get_blob_fields`

File: dagshub/data_engine/model/query_result.py [hunk 2]
```diff
@@ -408,9 +431,14 @@ def get_annotations(self, **kwargs) -> "QueryResult":
...
         """
         if len(self.annotation_fields) == 0:
-            logger.warning("No annotation fields in this query result")
             return self
-        return self.get_blob_fields(*self.annotation_fields, **kwargs)
+
+        log_message(f"Downloading annotation fields {self.annotation_fields}...")
+        if "load_into_memory" in kwargs:
+            del kwargs["load_into_memory"]
+        self.get_blob_fields(*self.annotation_fields, load_into_memory=True, **kwargs)
```
[the remark this answers is not in the document]
Comment (kbolashev) [reply dagshub/data_engine/model/query_result.py:439]: No, however I pass the kwargs to the downloading functions, and loading annotations requires them to be already in memory, so `load_into_memory=False` will break things.
Actually I should probably handle this particular edge case better in `get_blob_fields` and either skip loading if `load_into_memory=False` (explicitly respecting the user, even if it doesn't make sense), or force-opening.

File: dagshub/data_engine/model/query_result.py [hunk 3]
```diff
@@ -380,6 +382,27 @@ def _get_blob_fn(dp: Datapoint, field: str, url: str, blob_path: Path):
                         logger.warning(f"Got exception {type(exc)} while downloading blob: {exc}")
                     progress.update(task, advance=1)
 
+        # Convert any downloaded annotation column
```
[the remark this answers is not in the document]
Comment (kbolashev) [reply dagshub/data_engine/model/query_result.py:391]: Done, added to the docstring

File: dagshub/data_engine/annotation/metadata.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        annotation_type: Literal["bbox", "segmentation", "pose"],
+        annotation: Union[str, "ultralytics.engine.results.Results"],
+        yolo_context: Optional["YoloContext"] = None,
+        image_width: Optional[int] = None,
+        image_height: Optional[int] = None,
+    ):
+        """
+        Add a YOLO annotation.
+
+        This could be either a string of an annotations from a YOLO file, or a result of evaluating a YOLO model.
```
[the remark this answers is not in the document]
Comment (kbolashev) [reply dagshub/data_engine/annotation/metadata.py:266]: Changed it to a categories dict instead

File: dagshub/data_engine/annotation/metadata.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+            ann.height = bbox_height
+
+        self.annotations.append(ann)
+        self._update_datapoint()
+
+    def add_yolo_annotation(
+        self,
+        annotation_type: Literal["bbox", "segmentation", "pose"],
+        annotation: Union[str, "ultralytics.engine.results.Results"],
+        yolo_context: Optional["YoloContext"] = None,
```
[the remark this answers is not in the document]
Comment (kbolashev) [reply dagshub/data_engine/annotation/metadata.py]: Yep, did this in the end

**Pull request merged**